### PR TITLE
Return send mail promise

### LINKED ISSUE #21740: Mail service doesn't return driver/transport result anymore
### Describe the Bug

[Since 10.10.0](https://github.com/directus/directus/pull/21683) the result of an email transport is not being return anymore. There is no option to get this result, while we have production services relying on this result.

The usecase: we send out mails with Directus and store the id of the transport email message so we can pickup webhooks later on those messages.

@paescuj asked me to create this issue.

### To Reproduce

n/a

### Directus Version

v10.10.1

### Hosting Strategy

Self-Hosted (Docker Image)

### PR DESCRIPTION
**Pull request by appy-one**
## Scope

What's changed:

- This returns the Promise returned by `this.mailer.sendMail` so callers can optionally `await` and use the result, such as the ID returned by Mailgun

## Potential Risks / Drawbacks

- None

## Review Notes / Questions

- This ~partially~ reverts the changes made in #21683 ~, but still does not `await` the result (which was the goal of that PR)~

- Removed `await`s on calls to `mailService.send()` calls in the codebase, added `.catch()` handlers to log any thrown errors

---

Fixes #21740

### CODE REVIEW

**Comment by nickrum** [comment]
Since awaiting a promise in an async function and returning the result is equivalent to returning the promise directly, this effectively fully reverts the linked PR.

**Comment by appy-one** [comment]
> Since awaiting a promise in an async function and returning the result is equivalent to returning the promise directly, this effectively fully reverts the linked PR.

No - if you return the promise without `await` is will return an unresolved Promise without waiting for it to resolve first

**Comment by br41nslug** [comment]
> No - if you return the promise without await it will return an unresolved Promise without waiting for it to resolve first

Due to how async functions and promises work, both `async () => new Promise()` and `async () => await new Promise()` behave the same. Chaining promises will collapse them into a single promise so returning the promise in the async function results in identical behavior as before the linked PR

**Comment by appy-one** [comment]
> > No - if you return the promise without await it will return an unresolved Promise without waiting for it to resolve first
> 
> Due to how async functions and promises work, both `async () => new Promise()` and `async () => await new Promise()` behave the same. Chaining promises will collapse them into a single promise so returning the promise in the async function results in identical behavior as before the linked PR

Sorry @br41nslug you are wrong. If you return a Promise in an `async` function, you will not be able to catch errors anymore either using a `try` / `catch` block because the Promise will resolve outside of the scope of the `async` function. The `async` function will not `await` a returned promise unless you do so with an `await`.

**Comment by br41nslug** [comment]
> The async function will not await a returned promise unless you do so with an await.

Correct error handling is slightly different in those cases. The behavior is still identical, as you mention, when awaiting or then'ing the parent async function (which i should have explicitly mentioned but assumed that was implied as we're talking about getting the response back from the send function)

**Comment by appy-one** [comment]
To conclude, the code in this PR allows to receive the result of the `this.mailer.sendMail` for callers `await`ing the `mailService.send()` call. If one does not want to wait for the mail to be sent, they should not `await` the call, like @keesvanbemmel suggested. Error handling should be added by those callers however using `.catch(...)`

**Comment by br41nslug** [comment]
> If one does not want to wait for the mail to be sent, they should not await the call, like keesvanbemmel suggested. Error handling should be added by those callers however using .catch(...)

I believe the original PR was aiming to solve this issue for Directus built-in emails where the end user is not explicitly in control of whether it is awaited or not. If true then it was likely just fixed on the wrong level.

**Comment by appy-one** [comment]
Shall I adjust the current `await mailService.send()` calls in the codebase?

**Comment by rijkvanzanten** [comment]
@appy-one Yes please! 🙌🏻

File: api/src/operations/mail/index.ts [hunk 1]
```diff
@@ -29,6 +32,8 @@ export default defineOperationApi<Options>({
 			mailObject.html = type === 'wysiwyg' ? safeBody : md(safeBody);
 		}
 
-		await mailService.send(mailObject);
+		mailService.send(mailObject).catch((error: any) => {
+			logger.error(error.message);
+		});
```
Comment (paescuj) [inline api/src/operations/mail/index.ts:37]:
```suggestion
		mailService.send(mailObject).catch((error) => {
			logger.error(error, 'Could not send mail in "mail" operation');
		});
```

File: api/src/services/notifications.ts [hunk 2]
```diff
@@ -52,18 +52,18 @@ export class NotificationsService extends ItemsService {
...
+					.send({
 						template: {
 							name: 'base',
 							data: user['role']?.app_access ? { url: manageUserAccountUrl, html } : { html },
 						},
 						to: user['email'],
 						subject: data.subject,
+					})
+					.catch((error: any) => {
+						logger.error(`Could not send email`, error);
```
Comment (paescuj) [inline api/src/services/notifications.ts:65]:
```suggestion
					.catch((error) => {
						logger.error(error, `Could not send notification via mail`);
```

File: api/src/services/shares.ts [hunk 3]
```diff
@@ -157,16 +159,20 @@ ${userName(userInfo)} has invited you to view an item in ${share['collection']}.
...
-				},
-				to: email,
-				subject: `${userName(userInfo)} has shared an item with you`,
-			});
+					to: email,
+					subject: `${userName(userInfo)} has shared an item with you`,
+				})
+				.catch((error: any) => {
+					logger.error(`Could not send email`, error);
+				});
```
Comment (paescuj) [inline api/src/services/shares.ts:175]:
```suggestion
				.catch((error) => {
					logger.error(error, `Could not send share notification mail`);
				});
```

File: api/src/operations/mail/index.ts [hunk 4]
```diff
@@ -12,6 +13,8 @@ export type Options = {
 	subject: string;
 };
 
+const logger = useLogger();
```
Comment (paescuj) [inline api/src/operations/mail/index.ts:16]: @rijkvanzanten Do you prefer to have those calls on top-level vs. inside functions where it is used?

File: api/src/services/users.ts [hunk 5]
```diff
@@ -406,17 +408,21 @@ export class UsersService extends ItemsService {
...
+							data: {
+								url: this.inviteUrl(user?.email ?? email, url),
+								email: user?.email ?? email,
+							},
 						},
-					},
-				});
+					})
+					.catch((error: any) => {
+						logger.error(`Could not send email`, error);
```
Comment (paescuj) [inline api/src/services/users.ts:424]:
```suggestion
					.catch((error) => {
						logger.error(error, `Could not send user invitation mail`);
```

File: api/src/services/users.ts [hunk 6]
```diff
@@ -474,17 +480,21 @@ export class UsersService extends ItemsService {
...
+						url: acceptURL,
+						email: user.email,
+					},
 				},
-			},
-		});
+			})
+			.catch((error: any) => {
+				logger.error(`Could not send email`, error);
+			});
```
Comment (paescuj) [inline api/src/services/users.ts:497]:
```suggestion
			.catch((error) => {
				logger.error(error, `Could not send password reset mail`);
			});
```

**Review by paescuj — COMMENTED** [review]
Thanks @appy-one ❤️ Just added a few smaller notes, but looks good to me in general!

File: api/src/operations/mail/index.ts (hunk 4 above)
Comment (rijkvanzanten) [reply api/src/operations/mail/index.ts:16]: Either is fine for now 👍🏻

**Review by rijkvanzanten — APPROVED** [review]
Code changes LGTM!

**Comment by paescuj** [comment]
@appy-one Since I don't have write access to your branch, may I ask you to add the following changeset?

```md
---
'@directus/api': patch
---

Updated the `send` method in `MailService` to return the promise with info, adapted use within `@directus/api` to not block until mails have been sent
```

**Pull request merged**